File: specs/_estrutura_modulos/bindings/python/_template/api/src/logger.py
"""Logger estruturado do modulo <modulo>. Lei dona: specs/arquitetura/03-operacao.md §3.

Uma linha JSON por evento, com requestId. Campos de `camposSensiveis` sao redigidos AQUI — nao e
responsabilidade de quem chama lembrar. `print()` e proibido no modulo (regra `log`): a saida vai
por `sys.stdout`, que e o unico canal do logger.
"""
from __future__ import annotations

import json
import sys
from typing import Any, Sequence

NIVEIS = ("debug", "info", "warn", "error")
REDIGIDO = "[REDIGIDO]"
_PROFUNDIDADE_MAXIMA = 4


def _redigir(valor: Any, sensiveis: set[str], profundidade: int = 0) -> Any:
    """Substitui recursivamente o valor de todo campo sensivel."""
    if profundidade > _PROFUNDIDADE_MAXIMA:
        return valor
    if isinstance(valor, list):
        return [_redigir(item, sensiveis, profundidade + 1) for item in valor]
    if not isinstance(valor, dict):
        return valor
    return {
        chave: REDIGIDO if chave in sensiveis else _redigir(conteudo, sensiveis, profundidade + 1)
        for chave, conteudo in valor.items()
    }


class Logger:
    def __init__(self, modulo: str, nivel_minimo: str, campos_sensiveis: Sequence[str]) -> None:
        self._modulo = modulo
        self._minimo = NIVEIS.index(nivel_minimo)
        self._sensiveis = set(campos_sensiveis)

    def _emitir(self, nivel: str, mensagem: str, dados: dict[str, Any] | None) -> None:
        if NIVEIS.index(nivel) < self._minimo:
            return
        linha = {"nivel": nivel, "modulo": self._modulo, "mensagem": mensagem}
        linha.update(_redigir(dados or {}, self._sensiveis))
        sys.stdout.write(f"{json.dumps(linha, ensure_ascii=False)}\n")

    # `dados` e KEYWORD-ONLY de proposito: alem de deixar a chamada legivel, impede que o linter
    # confunda este logger com o `logging` da stdlib, cujo segundo argumento posicional e
    # argumento de format string. Sem isso, `logger.error("msg", {...})` vira falso positivo.
    def debug(self, mensagem: str, *, dados: dict[str, Any] | None = None) -> None:
        self._emitir("debug", mensagem, dados)

    def info(self, mensagem: str, *, dados: dict[str, Any] | None = None) -> None:
        self._emitir("info", mensagem, dados)

    def warn(self, mensagem: str, *, dados: dict[str, Any] | None = None) -> None:
        self._emitir("warn", mensagem, dados)

    def error(self, mensagem: str, *, dados: dict[str, Any] | None = None) -> None:
        self._emitir("error", mensagem, dados)


def criar_logger(modulo: str, nivel_minimo: str, campos_sensiveis: Sequence[str]) -> Logger:
    """`nivel_minimo` vem de config/api.json:nivelLog — nunca literal no codigo."""
    return Logger(modulo, nivel_minimo, campos_sensiveis)

File: specs/_estrutura_modulos/bindings/javascript/_template/api/src/logger.js
// Logger estruturado do modulo <modulo>. Lei dona: specs/arquitetura/03-operacao.md §3.
//
// Uma linha JSON por evento, com requestId. Campos de `camposSensiveis` sao redigidos AQUI —
// nao e responsabilidade de quem chama lembrar. `console.*` e proibido no modulo (regra `log`);
// a saida vai por process.stdout, que e o unico canal do logger.

const NIVEIS = ['debug', 'info', 'warn', 'error'];
const REDIGIDO = '[REDIGIDO]';
const PROFUNDIDADE_MAXIMA = 4;

/** Substitui recursivamente o valor de todo campo sensivel. */
function redact(valor, sensiveis, profundidade = 0) {
  if (profundidade > PROFUNDIDADE_MAXIMA || valor === null || typeof valor !== 'object') return valor;
  if (Array.isArray(valor)) return valor.map((item) => redact(item, sensiveis, profundidade + 1));

  const saida = {};
  for (const [chave, conteudo] of Object.entries(valor)) {
    saida[chave] = sensiveis.has(chave) ? REDIGIDO : redact(conteudo, sensiveis, profundidade + 1);
  }
  return saida;
}

/**
 * @param {{ modulo: string, nivelMinimo: string, camposSensiveis: string[] }} opcoes
 */
export function createLogger({ modulo, nivelMinimo, camposSensiveis }) {
  const sensiveis = new Set(camposSensiveis);
  const minimo = NIVEIS.indexOf(nivelMinimo);

  const emit = (nivel, mensagem, dados) => {
    if (NIVEIS.indexOf(nivel) < minimo) return;
    const linha = { nivel, modulo, mensagem, ...redact(dados ?? {}, sensiveis) };
    process.stdout.write(`${JSON.stringify(linha)}\n`);
  };

  return {
    debug: (mensagem, dados) => emit('debug', mensagem, dados),
    info: (mensagem, dados) => emit('info', mensagem, dados),
    warn: (mensagem, dados) => emit('warn', mensagem, dados),
    error: (mensagem, dados) => emit('error', mensagem, dados),
  };
}
